### ISSUE TITLE
Incremental Components in rasa_nlu

### ISSUE THREAD

**Issue by anrafla** [RasaHQ/rasa_nlu#1773]
I'm currently a master’s student currently working on a significant new feature to rasa_nlu: incremental components. The theory behind this feature can be found at: 

http://www.aclweb.org/anthology/W16-3631 and
 https://journals.linguisticsociety.org/elanguage/dad/article/download/361/361-2892-1-PB.pdf

While these papers are fairly dense, the main high-level idea is, while rasa_nlu components currently require the entire message (or 'utterance') for each step in the pipeline, an incremental component would work on the word level instead. The data is passed around components as incremental units. Incremental units can be added or revoked, with an “add” indicating new data to be processed by the nlu component, and “revoke” indicating that previous information was incorrect, and to disregard the information. These IUs are represented in python as tuples; (word, “add”) or (word, “revoke”). For example, take the query “Find me a Thai restaurant nearby”. With current functionality, the entire sentence is passed through the pipeline, and the intents and entities are returned at the end. With incremental components, as each word comes in, the components begin processing with the information they have. The above query would be represented as

(“Find”, “add”), (“me”, “add”) (“a”, “add”), (“Thai”, “add”), (“restaurant”, “add”), (“nearby”, “add”).
 
with each new IU, the most likely intent and entities are available in rasa_nlu’s Message object. This lays the groundwork for a more interactive dialogue system down the line, as information can be processed and made available before the entire utterance is finished. This can be used with already existing DM’s for processing to complete sooner, and a more interactive user-experience.

The changes I’m proposing are meant to be non-invasive and not affect or change any current functionality of the system. I have also implemented an example incremental component included, which isn’t particularly accurate, but provides an example workflow until future incremental components are created.

**Comment by akelad** [comment]
This sounds like an interesting idea, glad you're working with rasa_nlu during your masters!
Could I ask for some more clarification on when this would be useful though? As far as I understand it, it's about cutting down processing time?

**Comment by anrafla** [comment]
In essence, yes. The overall processing time may take longer because all the information isn't immediately available, but because the processing begins at the first word, it finishes _sooner_ than a non-incremental counterpart. Furthermore, an incremental NLU can be very powerful when coupled with an incremental DM, as it allows the system to be more interactive, i.e mid-utterance clarification requests, input affirmation, and filled pauses. It allows for an overall more user-friendly conversational experience, rather than a "ping-pong" style conversation. Here's a paper that explains how this style of interaction is preferred: http://www.aclweb.org/anthology/W13-4042 

Hopefully this cleared things up a bit, if not, I can go into further detail or try to explain a different way.

**Comment by akelad** [comment]
Definitely sounds like an interesting idea -- not sure we'll be incorporating something like this into our repository (it would require some discussion), but do you have the code available publicly somewhere so we can take a look?

**Comment by anrafla** [comment]
Yes - I just forked it to my research groups github and updated it with my changes so far 
https://github.com/bsu-slim/rasa_nlu

The main changes are:
Incremental_Interpreter in model.py, for incremental components being used by other python applications, and to preserve command-line evaluation without changes
Incremental_Component in component.py, adding a required new_utterance() method which is inherently required for an incremental system
--incremental command line flag in test.py, for command-line evaluation with Incremental_Interpreter

There is also rasa_nlu/sium.py and sium/sium_component.py, which are just for our sample incremental component, We just haven't gotten around to moving them in the right places.

Keep in mind the work is still not complete, the main functionality is there, but we have more testing/documentation to add.

**Comment by akelad** [comment]
awesome, we'll definitely take a look :)

**Comment by wochinge** [comment]
@Ghostvv @tabergma @dakshvar22 @JEM-Mosig What are thoughts on this one?

**Comment by Ghostvv** [comment]
@anrafla interesting idea, what algorithm do you use to extract entities from your your incremental components?

**Comment by Ghostvv** [comment]
@anrafla did you manage to implement your idea. As I see it, it requires significant changes to Rasa NLU and Core. Especially incremental dialogue manager.

**Comment by anrafla** [comment]
@Ghostvv The features was completed back in Q2 2019 and I believe some work has been put into it since I've finished -- the fork of the project is located here - https://github.com/bsu-slim/rasa_nlu

There's a short paper if you want a high level view of the changes we made - https://arxiv.org/pdf/1907.05403v1.pdf 

I can reach out to the current maintainer and put him in contact with your team.
